suite/tests/linux/syscall_pwait.c

```diff
@@ -109,6 +117,7 @@ main(int argc, char *argv[])
     sigdelset(&test_set, SIGUSR1);
     sigdelset(&test_set, SIGUSR2);
 
+    *ready_to_listen = false;
```

Comment:
Are you assuming the fork syscall is a barrier on arm to ensure this write is visible to the child before it hits its while loop?  Could you point at documentation for that?  Maybe an explicit barrier should be added.  Or switch to sysv ipc or sthg.  Ditto below.

Reply:
Yes, weak memory ordering, and yes 1 followed by 0 can happen.  Why do you say "I don't think this applies here"?  Because of the fork system call being a barrier?  Please be explicit.  I think any un-synchronized thread-shared store in code run on ARM needs a comment explaining why it is safe.  We need to be careful here: we can't be fast and loose with un-synchronized memory stores in cross-platform code.  We need to get the x86 strong ordering model out of our heads.  It has burned us in the past.  See #2502 where we still have outstanding bugs.

Reply:
But by the time *ready_to_listen = false (causality), the second process doesn't exist yet. According to fork, at the time of creation, both memory spaces have the same contents at the moment of the fork (in the program). The kernel must insert barriers there for ARM. So the "false" value is guaranteed. I can insert extra barriers but I strongly think this isn't necessary. Agree?

Reply:
A race detection tool will call this a data race.  The stores setting ready_to_listen to true rely on the blocking syscall as a sort of barrier, but there are no comments about that.  There are subtle issues here and I don't think we should be using lock-free communication where it's not hard to use safer synchronized communication (i.e., this isn't a perf-critical part of core DR; it's not assembly code: it's high-level code in a test).  We have suite/tests/condvar.h for safe thread communication; sthg similar should be used for process communication.  Or, the test could be tweaked to use threads, pthread_kill, and the existing condvar.h condition variables.  Most other projects have style/content guidelines that strongly discourage any kind of hand-rolled synch such as lock-free communication where at all possible.  I know that existing DR code is not always a paragon of best high-level practices but we should try to improve for better ARM support and cleaner code.

Looking at the test more closely: it seems inherently racy for SIGUSR2.  SIGUSR2 is not blocked up front, so it will be delivered immediately.  So if the child sends it before the parent executes its blocking syscall, the signal will go through and the parent will subsequently block.  The SIGUSR1 will interrupt it, but the parent will loop and block again and hang forever because the child does not send any further signals.  Right?  And the changes here don't eliminate that race: they just narrow the window for SIGUSR2 coming through too early.  Right?  Unless I'm missing something, then, I don't see how this PR fixes the race?  I think the only fix is to block SIGUSR2 up front as well.  Or else change the child to keep sending signals in a loop until both syscalls are interrupted.

Looking at the #3336 test output: could you explain your theory as to what is happening?  How is SIGUSR1 not being delivered at all?  The problem I see, described above, is that the test might hang.

Looking at the code for handle_pre_extended_syscall_sigmasks(): it fails to call check_signals_pending() so it won't deliver a pending signal like handle_sigprocmask() will.  Is this why SIGUSR1 is not being delivered?  That call should be added.

Reply:
Updated patch checks for pending signals. One complication was that we are changing the signal mask in pre_system_call. In sigreturn handler we re-set the app's emulated signal mask to the one in the signal frame. I have added code such that we are restoring the correct signal mask that we had previously saved in the "pre" part of a system call. This is already tested in syscall_pwait by the "sigmask mismatch" checks.